apps/src/javaide/JavaConsole.jsx

```diff
@@ -0,0 +1,152 @@
+import React from 'react';
+import {connect} from 'react-redux';
+import PropTypes from 'prop-types';
+import {appendInputLog} from './redux';
+import CommandHistory from '@cdo/apps/lib/tools/jsdebugger/CommandHistory';
+import {KeyCodes} from '@cdo/apps/constants';
+
+const style = {
+  consoleStyle: {
+    backgroundColor: 'black',
+    color: 'white',
+    height: '200px',
+    overflowY: 'auto'
+  },
+  consoleLogs: {
+    overflow: 'auto',
+    lineHeight: 'normal',
+    cursor: 'text',
+    whiteSpace: 'pre-wrap',
+    flexGrow: 1
+  },
+  consoleInputWrapper: {
+    flexGrow: 0,
+    flexShrink: 0,
+    display: 'flex'
+  },
+  consoleInputPrompt: {
+    display: 'block',
+    width: 15,
+    cursor: 'text',
+    flexGrow: 0
+  },
+  consoleInput: {
+    flexGrow: 1,
+    marginBottom: 0,
+    boxShadow: 'none',
+    backgroundColor: 'black',
+    color: 'white',
+    border: 'none'
+  }
+};
+
+/**
+ * Set the cursor position to the end of the text content in a div element.
+ * @see http://stackoverflow.com/a/6249440/5000129
+ * @param {!HTMLDivElement} element
+ */
+function moveCaretToEndOfDiv(element) {
+  const range = document.createRange();
+  if (element.childNodes.length === 0) {
+    return;
+  }
+
+  range.setStart(element.lastChild, element.lastChild.nodeValue.length);
+  range.collapse(true);
+
+  // Change window selection to new range to set cursor position
+  const selection = window.getSelection();
+  selection.removeAllRanges();
+  selection.addRange(range);
+}
+
+class JavaConsole extends React.Component {
+  static propTypes = {
+    // populated by redux
+    consoleLogs: PropTypes.array,
+    appendInputLog: PropTypes.func
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      commandHistory: new CommandHistory()
+    };
+  }
+
+  componentDidUpdate(prevProps) {
+    const prevLogsLength = prevProps.consoleLogs.length;
+    if (
+      typeof prevLogsLength === 'number' &&
+      prevLogsLength !== this.props.consoleLogs.length
+    ) {
+      this.jumpToBottom();
+    }
+  }
+
+  jumpToBottom = () => {
+    this._consoleLogs.scrollTop = this._consoleLogs.scrollHeight;
+  };
+
+  displayConsoleLogs() {
+    return this.props.consoleLogs.map((log, i) => {
+      let prefix = '<';
+      if (log.type === 'input') {
+        prefix = '>';
+      }
+      return (
+        <p key={`log_${i}`}>
+          {prefix} {log.text}
+        </p>
+      );
+    });
+  }
+
+  onInputKeyDown = e => {
+    const input = e.target.value;
+    if (e.keyCode === KeyCodes.ENTER) {
+      e.preventDefault();
+      e.target.value = '';
+      this.state.commandHistory.push(input);
+      this.props.appendInputLog(input);
+    } else if (e.keyCode === KeyCodes.UP) {
```

Comment:
@moneppo - Thinking about discoverability & touch
Is this an area we'd want to add in a up/down button to duplicate behavior, or is copying/pasting previous commands good enough as an alternate functionality?

Reply:
Great question. I need to think on this a bit, but my initial gut based on our discussion today is that devices without keyboards (e.g. phones, tablets without connected keyboards) are meant primarily for viewing the project. If a user really wants to enter a command, then yea, copy/paste is sufficient for repetition.